### PR TITLE
Speedup in case of multiple hits

### PR DESCRIPTION
**Pull request by fazouane-marouane**
This PR reimplements the lru-queue using a ES2015 Set. See discussion in https://github.com/medikoo/lru-queue/issues/2

It may need a polyfill for `Set` for people that need to use it on platforms that don't support it.

### CODE REVIEW

**Comment by medikoo** [comment]
@fazouane-marouane this looks very promising! Still I think we need to split it into two PR's, first one will concern current v0.1 release branch, and second to be issued as v1.

In v0.1 we can't jump to using native `Set`, and relying on polyfill has no chance to make it faster, than relying on `Object.create` directly. Therefore I propose that in this branch you'll propose fastest version that works with available ES5 means (as that we need to support across v0.1). We'll publish it as v0.1.1

We'll merge that one first and then derive `v0.1` branch out of that (for eventual further v0.1 patches)

In further PR it'll be great if you propose version that relies directly on native `Set` (which we'll publish as v1.0). You may hold with this one, until we above one merged (as I'd like to update linter settings so we fully rely on ES2017 syntax in that case).

In the meantime I've upgraded linter within package. so we rely on eslint. It'll create extra conflicts for you, sorry about that, but maintenance wise I think it's better to have that sorted now

**Comment by fazouane-marouane** [comment]
@medikoo sweet! This is good news. I totally agree with you, and I'll split this into two PRs then, one with Set for 1.0 and the other non-breaking changes + small optimisations for the 0.1.x

I'll probably have some time tonight or tomorrow.

**Comment by fazouane-marouane** [comment]
@medikoo You can find now on this PR the benchmark code + optimisation for es5 only. It's still not as better as a linkedlist implementation which uses es6 Map or the simple implementation that uses es6 Set but it's still gives comparable or better results than the old implementation (See result of the benchmark below). The idea is to not change anything if we hit the last entry on the cache. I also fixed linting errors.

> ----randomInsert----
Old implementation -> 56042.81594499946ms
Current implementation -> **61205.85211099684ms**
Linked list implementation -> 12405.282076001167ms
----incrementInsert----
Old implementation -> 40589.99467301369ms
Current implementation -> **8107.116282001138ms**
Linked list implementation -> 2664.3686120063066ms
----frequentHit----
Old implementation -> 37231.26925599575ms
Current implementation -> **1898.0850259959698ms**
Linked list implementation -> 2115.8094889968634ms
----cyclical----
Old implementation -> 22715.59111098945ms
Current implementation -> **23120.613740995526ms**
Linked list implementation -> 6557.003205999732ms
----fibo_5----
Old implementation -> 2858.357014000416ms
Current implementation -> **2911.5520230084658ms**
Linked list implementation -> 1062.3178920000792ms
----fibo_1000----
Old implementation -> 1863.0691489875317ms
Current implementation -> **1839.7867639958858ms**
Linked list implementation -> 1259.6590899974108ms

**Comment by medikoo** [comment]
Thank you!

I think it'll be good if we just update current solution with fastest approach, and do not host here different implementations (that could be in scope of some other repository)

Benchmark may stay purely as a mean to test current solution (and internally should also use only ES5 means), also maybe for benchmarking it'll be better to use some external tool. that's proven to measure things reliably? (but It's also fine for me if we just rely on `performance.now` here.

What do you think?

**Comment by fazouane-marouane** [comment]
Hello @medikoo 
Since I'm a bit under water lately, and can't have the perfect PR, I decided splitting it into 3 PRs (sorry about the spam). This one has the minimal impact, since it's a small improvement without changing anything about the core algorithm. There are no benchmark in the PR, as well.

#4 is an alternative implementation is es5 which is faster, but I didn't know if you were interested in it for the v0.1.1 or a v0.2.0, or not. Since I won't be able to make heavy changes to this PR in the coming days, I preferred creating another PR if this is what you were expecting (fastest implementation in es5).

#5 contains my original proposition, lru queue using es6 Sets.

As for benchmarks, IMO it has added value only when we compare alternative implementations. That been said, I also agree that this may not be the best place for benchmark code.

**Comment by medikoo** [comment]
> Since I'm a bit under water lately, and can't have the perfect PR, I decided splitting it into 3 PRs (sorry about the spam).

It is fine. Thank you

> #4 is an alternative implementation is es5 which is faster, but I didn't know if you were interested in it for the v0.1.1 or a v0.2.0, 

As long as public API remains unchanged it can be at v0.1 (we want fastest there)

Anyway I think we may keep both at different branches, and one that's faster will be at `master`

> #5 contains my original proposition, lru queue using es6 Sets.

Ok that'll be candiate for v1.0 but after we sort out previous two.

Thank you!

**Comment by medikoo** [comment]
> here are no benchmark in the PR, as well.

Maybe we can add a benchmark (that tests purely main implementation) with other separate PR. It'll be good to have it confirm on speed differences

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "lru-queue",
-  "version": "0.1.0",
+  "version": "0.1.1",
```
Comment (medikoo) [inline package.json:3]: Let's not change versions

File: package.json [hunk 2]
```diff
@@ -25,8 +25,9 @@
     "root": true
   },
   "scripts": {
+    "fix-lint": "eslint --ignore-path=.gitignore . --fix",
```
Comment (medikoo) [inline package.json:28]: Let's not add this script

**Pull request closed without merging**